### PR TITLE
feat: add root user support

### PR DESCRIPTION
**Pull request by JonZeolla**
# Contributor Comments

This adds support for the `root` user at runtime, in parallel to `easy_infra`.  Below are the tests that I updated/considered:

Everyting run_tests calls needs user except container-global
- run_path_check - DONE
  - check_paths - DONE
- run_env - DONE
  - run_aws - DONE
  - run_azure - DONE
- run_tool - DONE
  - run_cloudformation - DONE
  - run_ansible - DONE
  - run_terraform - DONE
- global_tests - DONE
  - test_version_arguments - DONE
  - test_sh - DONE
- run_security - NOT NEEDED
- All of the exec_tests need to be passed a user - DONE
  - Set it to fail closed if this is forgotten.
- Fix the input of test to specify a user and not "all" - DONE

## Pull Request Checklist

Thank you for submitting a contribution to `easy_infra`!

In order to streamline the review of your contribution we ask that you review and comply with the below requirements:

- [X] Rebase your branch against the latest commit of the target branch
- [X] If you are adding a dependency, please explain how it was chosen
- [X] If manual testing is needed in order to validate the changes, provide a testing plan and the expected results
- [X] If there is an issue associated with your Pull Request, link the issue to the PR.

### CODE REVIEW

File: build/Dockerfile.base [hunk 1]
```diff
@@ -56,14 +56,18 @@ RUN groupadd --gid 53150 -r easy_infra \
                 git-codecommit.us-west-1.amazonaws.com \
                 git-codecommit.us-west-2.amazonaws.com \
     >> /home/easy_infra/.ssh/known_hosts \
- && apt-get remove -y ssh \
  && echo "# END preloaded known_hosts as of $(date)" >> /home/easy_infra/.ssh/known_hosts \
+ && apt-get remove -y ssh \
```
Comment (JonZeolla) [inline build/Dockerfile.base:60]: This file has the only meaningful changes everything else is testing related.

File: build/Dockerfile.base [hunk 2]
```diff
@@ -56,14 +56,16 @@ RUN groupadd --gid 53150 -r easy_infra \
...
+ && mkdir -p /root/.ssh \
+ && touch /root/.ssh/known_hosts \
+ && cp /home/easy_infra/.ssh/known_hosts /root/.ssh/known_hosts \
  && if [ "${TRACE}" = "true" ]; then \
     apt-get -y install --no-install-recommends libcap2-bin \
                                                strace \
   ; fi \
  && su easy_infra -c "mkdir -p /home/easy_infra/.local/bin/" \
- && echo "export PATH=/home/easy_infra/.local/bin:${PATH}" >> /home/easy_infra/.bashrc \
```
Comment (JonZeolla) [inline build/Dockerfile.base:66]: We are now using a venv instead of `--user` pip installs

**Comment by digeseiso** [comment]
> # Contributor Comments
> This adds support for the `root` user at runtime, in parallel to `easy_infra`. Below are the tests that I updated/considered:
> 
> Everyting run_tests calls needs user except container-global
> 
> * run_path_check - DONE
>   
>   * check_paths - DONE
> * run_env - DONE
>   
>   * run_aws - DONE
>   * run_azure - DONE
> * run_tool - DONE
>   
>   * run_cloudformation - DONE
>   * run_ansible - DONE
>   * run_terraform - DONE
> * global_tests - DONE
>   
>   * test_version_arguments - DONE
>   * test_sh - DONE
> * run_security - NOT NEEDED
> * All of the exec_tests need to be passed a user - DONE
>   
>   * Set it to fail closed if this is forgotten.
> * Fix the input of test to specify a user and not "all" - DONE
> 
> ## Pull Request Checklist
> Thank you for submitting a contribution to `easy_infra`!
> 
> In order to streamline the review of your contribution we ask that you review and comply with the below requirements:
> 
> * [x]  Rebase your branch against the latest commit of the target branch
> * [x]  If you are adding a dependency, please explain how it was chosen
> * [x]  If manual testing is needed in order to validate the changes, provide a testing plan and the expected results
> * [x]  If there is an issue associated with your Pull Request, link the issue to the PR.

@JonZeolla i think we should add some comments on why we decided to add support. Also i think we should update the docs to advise the user to run as root if they need to use easy_infra in github actions

File: easy_infra/utils.py [hunk 3]
```diff
@@ -258,6 +286,19 @@ def gather_tools_and_environments(
...
+    """Return a list of users, based on the simplified provided user"""
+    if user == "all":
+        users: list[str] = constants.USERS
+    elif user not in constants.USERS:
+        LOG.error(f"{user} is not a supported user, exiting...")
+        sys.exit(1)
+    else:
+        users: list[str] = [user]
+
+    return users
```
Comment (digeseiso) [inline easy_infra/utils.py:299]: Wouldnt it be better to fail immediately if the provided user does not exist?
```suggestion
def gather_users(*, user: str) -> list[str]:
    """Return a list of users, based on the simplified provided user"""
    if user not in constants.USERS:
         LOG.error(f"{user} is not a supported user, exiting...")
         sys.exit(1)
     elif user == "all":
           users: list[str] = constants.USERS
      else:
            users: list[str] = [user]
```

**Review by digeseiso — CHANGES_REQUESTED** [review]
@JonZeolla i think we should add some comments on why we decided to add support for the root user. Also i think we should update the docs to advise users to run as root if they need to use easy_infra in github actions and perhaps link the documents show that running as a non-root user is not supported? or maybe simply mentioning it is enough?.

The [contributing.md](https://github.com/SeisoLLC/easy_infra/blob/80c6e31a8015d50b6a5812a5d8cc6dc1de018ee2/CONTRIBUTING.md?plain=1#L70) talks about ensuring that the user from inside the container can write to the host, i think we could also update that to somewhat reflect that a new root user is available and maybe mention github actions there? You decide

Even though it appears that containers run by default as the easy_infra user, do we also want to add a security disclaimer about running the container as root or just expect that the users will know?

**Comment by JonZeolla** [comment]
@digeseiso I only updated the tech details because while it works I don’t want people to potentially use it and make an insecure environment. Instead I plan to point people to our GitHub action, etc for those use cases.

**Comment by JonZeolla** [comment]
@digeseiso added a mention in the tech docs on why we added the root user support, but that you should prefer the default still

**Pull request merged**